pkg/scheduler/internal/queue/scheduling_queue_test.go

```diff
@@ -1840,94 +1840,123 @@ func TestPerPodSchedulingMetrics(t *testing.T) {
 	pod := st.MakePod().Name("test-pod").Namespace("test-ns").UID("test-uid").Obj()
 	timestamp := time.Now()
 
-	// Case 1: A pod is created and scheduled after 1 attempt. The queue operations are
-	// Add -> Pop.
-	c := testingclock.NewFakeClock(timestamp)
 	logger, ctx := ktesting.NewTestContext(t)
 	ctx, cancel := context.WithCancel(ctx)
 	defer cancel()
-	queue := NewTestQueue(ctx, newDefaultQueueSort(), WithClock(c))
-	queue.Add(logger, pod)
-	pInfo, err := queue.Pop()
-	if err != nil {
-		t.Fatalf("Failed to pop a pod %v", err)
-	}
-	checkPerPodSchedulingMetrics("Attempt once", t, pInfo, 1, timestamp)
 
-	// Case 2: A pod is created and scheduled after 2 attempts. The queue operations are
-	// Add -> Pop -> AddUnschedulableIfNotPresent -> flushUnschedulablePodsLeftover -> Pop.
-	c = testingclock.NewFakeClock(timestamp)
-	queue = NewTestQueue(ctx, newDefaultQueueSort(), WithClock(c))
-	queue.Add(logger, pod)
-	pInfo, err = queue.Pop()
-	if err != nil {
-		t.Fatalf("Failed to pop a pod %v", err)
-	}
-	queue.AddUnschedulableIfNotPresent(logger, pInfo, 1)
-	// Override clock to exceed the DefaultPodMaxInUnschedulablePodsDuration so that unschedulable pods
-	// will be moved to activeQ
-	c.SetTime(timestamp.Add(DefaultPodMaxInUnschedulablePodsDuration + 1))
-	queue.flushUnschedulablePodsLeftover(logger)
-	pInfo, err = queue.Pop()
-	if err != nil {
-		t.Fatalf("Failed to pop a pod %v", err)
-	}
-	checkPerPodSchedulingMetrics("Attempt twice", t, pInfo, 2, timestamp)
+	tests := []struct {
+		name                            string
+		perPodSchedulingMetricsScenario func(*testingclock.FakeClock, *v1.Pod) (*framework.QueuedPodInfo, error)
+		wantAttempts                    int
+		wantInitialAttemptTs            time.Time
+	}{
+		{
+			// The queue operations are Add -> Pop.
+			name: "pod is created and scheduled after 1 attempt",
+			perPodSchedulingMetricsScenario: func(c *testingclock.FakeClock, pod *v1.Pod) (*framework.QueuedPodInfo, error) {
+				queue := NewTestQueue(ctx, newDefaultQueueSort(), WithClock(c))
+				queue.Add(logger, pod)
```

Comment:
can you please move this instantiation out into the test body and change `perPodSchedulingMetricsScenario` to receive queue as a parameter.

Reply:
Addressed.